### PR TITLE
Refactor BatchSpec type and ParseBatchSpec to be shareable

### PR DESCRIPTION
**Pull request by eseliger**

### CODE REVIEW

File: cmd/src/batch_common.go [hunk 1]
```diff
@@ -363,16 +364,22 @@ func executeBatchSpec(ctx context.Context, opts executeBatchSpecOpts) (err error
...
 	if err != nil {
 		return nil, "", err
 	}
 	defer f.Close()
 
-	return svc.ParseBatchSpec(f)
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, "", errors.Wrap(err, "reading batch spec")
+	}
```
Comment (eseliger) [inline cmd/src/batch_common.go:379]: don't care about a file flag in the backend, just wanna pass []byte or string

File: internal/batches/workspace/workspace_test.go [hunk 2]
```diff
@@ -64,30 +58,22 @@ func TestBestWorkspaceCreator(t *testing.T) {
 		},
 	} {
 		t.Run(name, func(t *testing.T) {
-			var steps []batches.Step
-			if tc.images != nil {
-				steps = make([]batches.Step, len(tc.images))
-				for i, image := range tc.images {
-					steps[i].SetImage(image)
-				}
-			}
```
Comment (eseliger) [inline internal/batches/workspace/workspace_test.go:73]: nice, IMO

File: internal/batches/executor/task_builder.go [hunk 3]
```diff
@@ -13,29 +13,18 @@ type DirectoryFinder interface {
...
-		conf.SetGlob(g)
-		tb.initializedWorkspaceConfigs = append(tb.initializedWorkspaceConfigs, conf)
-	}
-
-	return tb, nil
+// BuildTasks returns tasks for all the workspaces determined for the given spec.
+func BuildTasks(ctx context.Context, spec *batches.BatchSpec, finder DirectoryFinder, repos []*graphql.Repository) ([]*Task, error) {
+	tb := &taskBuilder{spec: spec, finder: finder}
+	return tb.buildAll(ctx, repos)
 }
```
Comment (eseliger) [inline internal/batches/executor/task_builder.go:25]: simpler API and doesn't need to store the glob on the steps

File: internal/batches/batch_spec.go [hunk 4]
```diff
@@ -151,9 +115,19 @@ type Group struct {
 	Repository string `json:"repository,omitempty" yaml:"repository"`
 }
 
-func ParseBatchSpec(data []byte, features FeatureFlags) (*BatchSpec, error) {
+type ParseBatchSpecOptions struct {
+	AllowArrayEnvironments bool
+	AllowTransformChanges  bool
+	AllowConditionalExec   bool
+}
```
Comment (eseliger) [inline internal/batches/batch_spec.go:122]: don't require featureFlags in backend, just make all these fields truthy

File: internal/batches/batch_spec.go [hunk 5]
```diff
@@ -95,8 +82,6 @@ type Step struct {
 	Outputs   Outputs           `json:"outputs,omitempty" yaml:"outputs,omitempty"`
 
 	If interface{} `json:"if,omitempty" yaml:"if,omitempty"`
-
-	image docker.Image
```
Comment (eseliger) [inline internal/batches/batch_spec.go:99]: don't want internal structures pollute this parse type

**Review by mrnugget — APPROVED** [review]
Approve as just discussed on Zoom.

Comments/names still need updating.

**Pull request merged**